File: bytedance/543.py
# 543. 二叉树的直径
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        self.ans = 1

        def depth(node: TreeNode):
            if not node: return 0
            l = depth(node.left)
            r = depth(node.right)
            self.ans = max(self.ans, l + r + 1) # 及时更新
            return max(l, r) + 1
        
        depth(root)
        return self.ans - 1

File: bytedance/543.java
class Solution {
    int ans;
    public int diameterOfBinaryTree(TreeNode root) {
        ans = 1;
        depth(root);
        return ans - 1;
    }

    public int depth(TreeNode node) {
        if (node == null) return 0;
        int l = depth(node.left)
        int r = depth(node.right)
        ans = Math.max(ans, l+r+1) // 更新最大直径
        return Math.max(l,r) + 1
    }
}